src/Microsoft.ComponentDetection.Detectors/pub/PubComponentDetector.cs

```diff
@@ -0,0 +1,68 @@
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.IO;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Microsoft.ComponentDetection.Contracts;
+using Microsoft.ComponentDetection.Contracts.Internal;
+using Microsoft.ComponentDetection.Contracts.TypedComponent;
+using YamlDotNet.Serialization;
+using YamlDotNet.Serialization.NamingConventions;
+
+namespace Microsoft.ComponentDetection.Detectors.Pub
+{
+    [Export(typeof(IComponentDetector))]
+    public class PubComponentDetector : FileComponentDetector
+    {
+        public override string Id { get; } = "Pub";
+
+        public override IEnumerable<string> Categories => new[] { Enum.GetName(typeof(DetectorClass), DetectorClass.Pub) };
+
+        public override IList<string> SearchPatterns { get; } = new List<string> { "pubspec.lock" };
+
+        public override IEnumerable<ComponentType> SupportedComponentTypes { get; } = new[] { ComponentType.Pub };
+
+        public override int Version { get; } = 1;
+
+        public PubComponentDetector()
+        {
+            NeedsAutomaticRootDependencyCalculation = true;
+        }
+
+        protected override Task OnFileFound(ProcessRequest processRequest, IDictionary<string, string> detectorArgs)
+        {
+            var singleFileComponentRecorder = processRequest.SingleFileComponentRecorder;
+            var file = processRequest.ComponentStream;
+
+            Logger.LogVerbose("Found Pubspec.lock: " + file.Location);
+            ParsePubspecYamlFile(singleFileComponentRecorder, file);
+
+            return Task.CompletedTask;
+        }
+
+        private void ParsePubspecYamlFile(ISingleFileComponentRecorder singleFileComponentRecorder, IComponentStream file)
+        {
+            var components = new Dictionary<string, DetectedComponent>();
+
+            var text = string.Empty;
+            using (var reader = new StreamReader(file.Stream))
+            {
+                text = reader.ReadToEnd();
+            }
+
+            var deserializer = new YamlDotNet.Serialization.DeserializerBuilder()
+                .WithNamingConvention(UnderscoredNamingConvention.Instance)
+                .Build();
+
+            dynamic pubspec = deserializer.Deserialize<System.Dynamic.ExpandoObject>(text);
```

Comment:
Can the deserialization contract be captured in a class like https://github.com/microsoft/component-detection/blob/main/src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmYaml.cs so we aren't using dynamic objects?

Reply:
I tried that in an earlier commit, and my testing was running into parsing failures because the lock files weren't consistent enough. The dynamic object seems to be more reliable, but I may have missed something. https://github.com/jhutchings1/component-detection/commit/829c548ffd19fedd6a3c294ad4cf59c410b15128